File: graph/build_map/chain_forward_star.cpp
// N为节点的数量，e, ne, w数组长度为边的数量
int h[N], e[N], ne[N], w[N], idx;
inline void add(int a, int b, int c)
{
	w[idx] = c, e[idx] = b, ne[idx] = h[a], h[a] = idx++;
}
// 记得初始化h
memset(h, -1, sizeof h);

File: graph/build_map/chain_forward_star.py
class Graph:
    def __init__(self, n, m):
        self.h = [-1] * (n + 10)
        self.e, self.ne, self.w = [0] * (m + 10), [0] * (m + 10), [0] * (m + 10)
        self.idx = 0

    def add(self, a, b, c):
        self.w[self.idx] = c; self.e[self.idx] = b
        self.ne[self.idx] = self.h[a]; self.h[a] = self.idx
        self.idx += 1
